src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -74,36 +74,61 @@ internal ODataPathParser(ODataUriParserConfiguration configuration)
         /// anything and simply provides the raw text of both the identifier and parenthetical expression.
         /// </summary>
         /// <remarks>Internal only so it can be called from tests. Should not be used outside <see cref="ODataPathParser"/>.</remarks>
+        /// <param name="enableKeyAsSegment">Whether key-as-segment is enabled.</param>
         /// <param name="segmentText">The segment text.</param>
         /// <param name="identifier">The identifier that was found.</param>
         /// <param name="parenthesisExpression">The query portion that was found. Will be null after the call if no query portion was present.</param>
-        internal static void ExtractSegmentIdentifierAndParenthesisExpression(string segmentText, out string identifier, out string parenthesisExpression)
+        internal static void ExtractSegmentIdentifierAndParenthesisExpression(
+            bool enableKeyAsSegment,
+            string segmentText,
+            out string identifier,
+            out string parenthesisExpression)
         {
             Debug.Assert(segmentText != null, "segment != null");
 
+            // We allow a single trailing '/', which results in an empty segment.
+            // However System.Uri removes it, so any empty segment we see is a 404 error.
+            if (segmentText.Length == 0)
+            {
+                throw ExceptionUtil.ResourceNotFoundError(ODataErrorStrings.RequestUriProcessor_EmptySegmentInRequestUrl);
+            }
+
             int parenthesisStart = segmentText.IndexOf('(');
-            if (parenthesisStart < 0)
+
+            bool containsOpenParenthesis = parenthesisStart >= 0;
+            bool endsWithCloseParenthesis = segmentText[segmentText.Length - 1] == ')';
+
+            if (containsOpenParenthesis != endsWithCloseParenthesis && !enableKeyAsSegment)
             {
-                identifier = segmentText;
-                parenthesisExpression = null;
+                // If the parenthesis are not in the expected state fail with a syntax error. We do not do this for key-as-segment
+                // because in that case the segment text is potentially still a valid key value.
+                throw ExceptionUtil.CreateSyntaxError();

```

Comment:
Is this condition consistent with existing behaviour?

In the current behaviour, if the segment does not contain and open parenthesis but ends with closing parenthesis, e.g. `folder)`, the condition would pass and the `identifier` would be set to the `segmentText`.

However, in your change, if the segment does not contain an open parenthesis but ends with closing parenthesis, the condition would fail when `enableKeyAsSegment` is false and a segment like `folder)` would throw an error.